packages/flutter_test/test/widget_tester_test.dart

```diff
@@ -666,6 +666,29 @@ void main() {
         key: 'abczed',
       });
     });
+    testWidgets('runs all microtasks within certain number of frames', (WidgetTester tester) async {
+      final Completer<int> completer = Completer<int>();
+      await tester.runAsync(() async {
+        Future<void>.delayed(const Duration(milliseconds: 150)).then((_) {
```

Comment:
This is scary. I do agree it will be convenient to have ability to wait a period of time, but this may be a source of flakiness.

I was actually thinking about solving this problem before, is it possible to get all the the timer and microtask created within a zone(maybe be caching when someone calls ZoneSpecification.scheduleMicrotask and ZoneSpecification.createPeriodicTimer), and provide some kinda of api to expose those?

Reply:
In general, I agree with you.

This PR does not make things less or more flaky. If your test relies on some condition to resolve asynchronously, then it may or may not get a chance to resolve when runAsync returns. This PR aims to make that *a little* better by trying to flush the microtask queue before returning. It makes sense but indeed, it is not a guarantee and thus it does not reduce flakes.

What it *does* do is to expose the pre-existing ugliness to developers. Now we have a ominous parameter `minFramesUntilIdle` which, if we want to, we can default to 0 so we don't wait at all (current behavior). With increased documentation and this new parameter, we are giving developers a strong signal that `runAsync` is working as intended and if they want less flakiness, they should consider changing their code.

Without this change, people are surprised by the behavior of runAsync and try to "fix" it by doing stuff like this:

```dart
runAsync(() => // some code that causes a side effect in my widgets);
runAsync(() => pumpEventQueue(times: someRandomNumber)); // I found this API that's supposed to help, I guess?
```

Fiddle with `someRandomNumber` until it passes. There's no other solution to this that I am aware of.

I am trying to make this part of the API and documentation so at least developers *hopefully* understand what they are doing and why they are doing it. If they start seeing flakes, they can refer to documentation and see why it is happening.

Replace code above with:

```dart
runAsync(
  () => // some code that causes a side effect in my widgets,
  minFramesTillIdle: someRandomNumber, // Ah, but now I see this is a synchronization issue.
);
```

As for exposing timers and microtasks, I think that's a problem for an IDE to solve. While debugging these tests, the users should be able to see what kind of timers and futures are now "frozen" because we ended the real async zone. I have a generic bug for this that I am hoping will get some traction. 

I tried to save this information and perhaps print a warning message that there are unresolved tasks and such but it is too noisy. It alerts when there's no need to alert (such as no side effect) and it doesn't alert when it needs to (e.g. some IO operation will eventually post a microtask but we have no evidence of that now).

The final thing we agree on: This problem bothers me :-). I think this PR makes things a bit more transparent (but, like you said, not any better).

Reply:
For exposing timer and microtasks, I meant expose a handle for the user to await for them. The problem is some of our code (image decoding for example) create futures without any reference for test to await for, and that is the main reason people want to do pumpEventQueue and hope those futures get resolved.

Reply:
It is not just Future's. Image has streams and other subscription mechanisms as well.  Is it possible to capture all those in a zone? I know that Future.delayed creates a Timer but I don't think an `ImageCompleter` stream would create anything that's observable in a zone. I will test this out and get back to you.

We would also need to create a polling mechanism with a completer to track timers and return when all timers return false for https://api.dart.dev/stable/2.8.4/dart-async/Timer/isActive.html as Timers themselves do not provide this functionality AFAIK.

Reply:
I believe the imageCompleter is just a wrapper over a code future, the future will either be a wrapper of timer or microtask, but then i think you are right that we need to find a way to flush the timer. flushing microtasks should be as easy as creates await Future.delay(0), but timer is tricky.

Reply:
Maybe things will be easier in the fake async zone that the runTest use, in there we can control the time flow so even if we do fakeAsync.flushTimer(Duration(hours:1)) it will be finished in a instant. we can expose all the timers and make a API that is something like ```await tester.resolveTimer(Timer timer)```

Reply:
I am not sure I understand. Futures/Timers cannot cross over zones. We cannot move a Timer that was created in real async zone to fake async and resolve it there.

However, I made some promising progress. If I keep track of timers, I can do this:

```dart
      final Completer<void> sideEffectCompleter = Completer<void>();
      const Duration checkInterval = Duration(milliseconds: 100);
      Zone.root.createPeriodicTimer(checkInterval, (Timer timer) {
        timerCache.retainWhere((Timer timer) => timer.isActive);
        repeatingTimerCache.retainWhere((Timer timer) => timer.isActive);
        if (microtasks == 0 && timerCache.isEmpty && repeatingTimerCache.isEmpty) {
          timer.cancel();
          sideEffectCompleter.complete();
        } else if ((checkInterval * timer.tick) >= additionalTime) {
          timer.cancel();
          sideEffectCompleter.completeError(TimeoutException('Timed out!'));
        }
      });
      try {
        final T value = await callback();
        // This is something we can expose to user or we can await ourselves.
        await sideEffectCompleter.future;
        return value;
      }
```
I want to build up a more comprehensive test involving streams. Let me know what you think.

Reply:
I think you meant exposing Timers/Futures etc that we use in the widgets for testing. Even so, there are certain things where we "have to" rely on real async such as file io. Widget tests need to be able to load images from assets which means a certain part of widget testing cannot be in fake async zone.

Reply:
I am not familiar with file io, why couldn't it run in fake async?  The assets loading part should be fine because we mocked the asset channel to read file synchronously https://github.com/flutter/flutter/blob/aa0382e95d3fe19b9ee4becf3fdf4017c15a149c/packages/flutter_test/lib/src/_binding_io.dart#L40

The reason I brought up fake async is that i am thinking we can move away from writing test in runAsync if we can make the fakeasync in testWidgets expose the timer and microtask.

Reply:
There has been numerous discussions about trying to run everything in fake async. See related issue: https://github.com/flutter/flutter/issues/23179 and specifically: https://github.com/flutter/flutter/issues/23179#issuecomment-452128558. For this to happen, we need to enumerate all the possible things that Flutter widgets might do and provide sync implementations of them (much like you did for file loading). I am not sure how feasible that is.

Even if it was feasible, do you think we can still make progress in making runAsync more robust and chase that problem separately? We should seek incremental and innovative solutions together. If we get rid of `runAsync` at some point, I would be *very* happy but it does not seem possible for the time being.

Reply:
I see yes you are right, I don't have an idea to fix the issue you mentioned neither. i agree we should make runAsync more robust at least for now.

Reply:
I updated the PR. I still had to introduce some "wait time" in there because microtasks can be scheduled by any async operation "back to back". However, I feel better about this approach because it is more deterministic and easy to explain. Let me know what you think.